src/ostree/ot-main.c

```diff
@@ -510,6 +511,11 @@ ostree_option_context_parse (GOptionContext *context, const GOptionEntry *main_e
 
   if (opt_version)
     {
+      /* Ignore SIGPIPE so that piping the output to grep or similar
+       * doesn't cause the process to fail. */
+      if (signal (SIGPIPE, SIG_IGN) == SIG_ERR)
```

Comment:
Actually wait though, shouldn't we be setting it to `SIG_DFL` here so we get killed by SIGPIPE, which the shell is expecting?

Reply:
No, that's exactly what's happening now. The default disposition for `SIGPIPE` is to kill the process. The process gets killed and the shell sets the exit status to 128 + signum = 141. The `pipefail` feature says that any process with a non-0 exit status is considered failed. I think it would be better if bash saw the `SIGPIPE` triggered exit and treated it as a successful exit, but that's not what it does. See https://lists.nongnu.org/archive/html/bug-bash/2017-03/msg00179.html for one of the util-linux maintainers unsuccessfully trying to convince the bash maintainers to do that.

By ignoring the signal, the process won't be killed. Instead, all the writes will fail with `EPIPE`, which `g_print` will ignore. That's not so nice since it will keep trying to write data even though there's no reader. However, since the `opt_version` output is small and controlled, it's no big deal.

Reply:
Right, thanks.  So basically there's no way to get strict error handling right now for pipes that works reliably.

Yeah.  I think the answer here is to not use shell script...

Reply:
> Yeah. I think the answer here is to not use shell script...

Exactly. For me it's kinda sad because my entry point to software was elaborate shell scripts to do various things. But after being bitten by one too many of these gotchas I've mostly stopped writing them unless they're very simple.